### ISSUE TITLE
Data persistence for API

### ISSUE THREAD

**Issue by AuHau** [codex-storage/nim-codex#885]
It's not the greatest title but I can't figure something better.

Currently, when you query endpoints like `/slots` or `/purchases` they are backed by the Marketplace's state-machines, and state-machines are loaded upon node's start from on-chain using the contract tracking `mySlots()` and `myRequests()`.

This mechanism does not really keep much of the data it represents around, mainly the Sales state-machine, because it cleans up after it finishes: https://github.com/codex-storage/nim-codex/blob/e6a387e8e81824b2fcd816f7bde25699bbe1a162/codex/sales.nim#L163

The purchasing machine I believe does not clean up so the "data" will stay there until the node shutdown, but then when it re-starts it won't "reload it" from the `myRequests()` because it is already gone.

This is not great API experience. Most probably for example UI will want to be able fetch history of purchases/sales to display to the user...

**Comment by emizzle** [comment]
I definitely agree this is not the best API experience and should be unified. One of the more obvious options is to keep past slots / requests in a mapping in the contract. That mapping could be returned in a `myPastSlots`/`myPastRequests` function in the contract and a REST API request to get past slots/requests could  simply return a `myPastSlots`/`myPastRequests`. The purchasing module would then remove past purchases from memory. That way, both purchasing and sales don't keep past purchases / sales agents in memory, and the REST API endpoints would get the data from the contract.

This would allow us to get rid of the notion of returning `PurchaseIds` in the REST API, and rely on the `RequestId` instead. This will be more consistent with how we retrieve sales info by using the `SlotId` and not a `SalesAgentId` (which doesn't exist). On node startup, we can retrieve `myPastRequests` and reverse-map it to a `PurchaseState`, similar to how it is currently done on `Purchasing.load`, except the mapping is reversed. Likewise for `myPastRequests` in the sales module. I never liked how the REST API returned a `PurchaseId` after creating a storage request. Most users expected it to return a `RequestId`. This even tripped up a lot of our team during runs we did for the EthCC workshop.

**Comment by AuHau** [comment]
Hmm, you are right that we already have the data in the contract. The question is how much we depend on the freeing gas refund of deleting the data from the mapping. If we are willing to take the hit, then it might be an interesting and more importantly simple solution. 

I am not sure if you propose to create another mapping for "past slots/request", if so, then I would not go that route. I would simply store it in the current mapping we have and have two view functions, one for fetching "active slots/requests" and one that would simply return all—I would not go the route of "fetching past events"—lets the API consumer filter out what he needs.

**Comment by emizzle** [comment]
Yeah that makes sense. Definitely keep it in the existing mapping. I don’t know if there will be consequences to storing data in a mapping that forever increases in size. Possibly the gas required to store data could also continually increase?

If that’s the case, the other option is to persist the data locally, which will not survive a catastrophic crash, of course.

Regardless, I still think we should attempt to remove PurchaseIds from the API flow if possible.

**Comment by AuHau** [comment]
> Possibly the gas required to store data could also continually increase?

I don't think so - there is one-time constant fee for storing something.

> If that’s the case, the other option is to persist the data locally, which will not survive a catastrophic crash, of course.

Honestly,  that would be my preferred

**Comment by emizzle** [comment]
Should we combine #458 with this task as well? Perhaps at first, we can do as we discussed in #458, and rebuild the state from event emissions. Then perhaps later we could add persistence in the repo store if needed?

**Comment by AuHau** [comment]
> I would also like to have past events stored locally, not on chain. We're already growing the on-chain state quite a bit, and adding more to it is going to hurt the scalability of the network. Ideally, I'd like to move towards a situation where we clean up each and bit of state related to a request as soon as all interactions with it are done.

I agree, and I am not suggesting the onchain approach to be the final design, but IMHO this should be part of the first release and I am not sure how much time we will have for proper implementation if we have the target that we have.

I would go for the first simple implementation with the chain state solution, and if we have time left, then redo it properly with a local-only approach. 

> ... rebuild the state from event emissions . Then perhaps later we could add persistence in the repo store if needed?

IMHO the "events rebuild" approach needs to be bundled with persistence from day 0. Because otherwise you need to fetch events from the moment that the contracts were deployed which can grow quickly. The validator's case is a bit special because we know that we need to fetch at most N blocks, which is defined by the request's duration limit we gonna have.

**Comment by emizzle** [comment]
> I would go for the first simple implementation with the chain state solution, and if we have time left, then redo it properly with a local-only approach.

I agree that the ease of this approach is going to be much more suitable given our timeline. We can treat the local-only solution as optimisation step later on when we have more time.

**Comment by emizzle** [comment]
> IMHO the "events rebuild" approach needs to be bundled with persistence from day 0. Because otherwise you need to fetch events from the moment that the contracts were deployed which can grow quickly.

New nodes would also need to sync the state from the initial block, yes. We discussed this already in #458, and this is why I argued against it initially (for the validator, without realising the 30 day duration window), however there are various things we could do like offload to another thread, store locally in repo store with block synced, etc. Were there other "local" approaches you were thinking of?
